### DISCUSSION TITLE
Is there a way to make WSL stop using a distro's filesystem file so that that distro can be exported?

### DISCUSSION THREAD

**Discussion by adamkpickering** [microsoft/WSL#10593] [Q&A]
Hello! I am a [Rancher Desktop](https://github.com/rancher-sandbox/rancher-desktop) developer. Similar to Docker Desktop, Rancher Desktop uses WSL on Windows to run Linux containers.

In our next release, we plan to include [a "snapshots" feature](https://github.com/rancher-sandbox/rancher-desktop/issues/4236). This will allow a user to take a snapshot of the state of Rancher Desktop, and restore to that snapshot later. In order to do this, Rancher Desktop uses `wsl.exe --export` and `wsl.exe --import`.

Of course, before exporting the distro that Rancher Desktop uses behind the scenes, that distro must not be in use. We currently achieve this by using `wsl.exe --terminate`. Prior to snapshots, this approach worked well. With snapshots, our approach looks like this:
1. Shut down the backend (which includes `wsl.exe --terminate`).
2. Create a snapshot (which includes `wsl.exe --export`).
3. Restart the backend.

But I am running into a problem. When I try to run `wsl.exe --export` immediately after shutting down the backend, I run into the following error:
```
The process cannot access the file because it is being used by another process.
Error code: Wsl/Service/0x80070020
```
From this, I understand that something is still using the distro's filesystem file (`ext4.vhdx`). In testing, I need to wait 50s-1m for whatever is using this file to stop using it. It also seems that a failed run of `wsl.exe --export` resets this process - I have to wait another 1m in order for it to work.

Is there any way of monitoring/expediting this process of giving up the file? I am finding that there is not a lot of documentation out there about this mechanism. Thanks!

**Comment by Yewzir** [comment]
@adamkpickering 

Instead of `wsl.exe --terminate`, I think what you need is `wsl.exe --shutdown`. The former command stops a distribution from running, or terminates it. The latter command, however, **forces** instant termination of all running distributions, continuously running background processes and the WSL 2 lightweight utility virtual machine.

Kind regards
Yewzir
